From: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>
Date: Fri, 5 Jan 2024 21:20:26 +0000
Subject: [PATCH 1/1] error.c: buffer overflow in msg2()

A buffer overrun in msg2() was observed when executing the HOSTNAMEOVFL
test.

If strncpy() truncates a string it does not append a terminating NUL
character. Insert a NUL character after the destination area.
This ensures that strchr() will stop at the end of the string.

Fixes: 9be423ceea3c ("Improved handling of very long host or program names, or no strftime")
Signed-off-by: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>

Forwarded: yes
Last-Update: 2024-01-05
---
 error.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/error.c b/error.c
index d0a2a9e..3435046 100644
--- a/error.c
+++ b/error.c
@@ -404,6 +404,7 @@ void msg2(
    if (bufp < buff+MSGLEN)
       *bufp++ = ' ';
    strncpy(bufp, text, MSGLEN-(bufp-buff));
+   bufp[MSGLEN-(bufp-buff)] = 0;
    bufp = strchr(bufp, '\0');
    strcpy(bufp, "\n");
    _msg(level, buff, syslp);
-- 
2.43.0

